lib/SILGen/SILGenThunk.cpp

```diff
@@ -159,10 +161,32 @@ SILGenModule::getOrCreateForeignAsyncCompletionHandlerImplFunction(
   std::copy(blockType->getParameters().begin(),
             blockType->getParameters().end(),
             std::back_inserter(implArgs));
+
+  const clang::Type *clangType = blockType->getClangTypeInfo().getType();
+  const clang::Type *newClangType = nullptr;
+  if (auto blockPtrTy = dyn_cast_or_null<clang::BlockPointerType>(clangType)) {
+    auto &clangCtx =
+        getASTContext().getClangModuleLoader()->getClangASTContext();
+    SmallVector<clang::QualType, 4> implClangArgs;
+    implClangArgs.push_back(
+        clang::QualType(getASTContext().getClangTypeForIRGen(blockStorageTy), 0));
+    if (auto fnProtoTy = dyn_cast_or_null<clang::FunctionProtoType>(
+                blockPtrTy->getPointeeType().getTypePtr())) {
+      llvm::copy(fnProtoTy->getParamTypes(), std::back_inserter(implClangArgs));
+      auto newFnTy = clangCtx.getFunctionType(
+          fnProtoTy->getReturnType(), implClangArgs,
+          fnProtoTy->getExtProtoInfo());
+      newClangType = clangCtx.getPointerType(newFnTy).getTypePtr();
+    } else {
+      llvm::report_fatal_error("Unexpected FunctionNoProtoType!");
```

Comment:
Did you test that this can't happen if there's a completion handler of type e.g. `void(^)()`?

Reply:
I've handled the case for now, but non-prototype completion handlers don't seem to be working. Will investigate/file a bug report.